elastic/elastic-agent
# make download of upgrade artifacts async

## What does this PR do?

upgrade artifacts are downloaded async when upgrade is initiated from Fleet.

- handle upgrade in separate go routine
- increase download timeout to 2 hours
- add KeepAliveSettings
- set idle connection timeout to 30 sec

## Why is it important?

Without this clients missed check-ins because they were busy downloading.

## Checklist

~~- [ ] I have made corresponding changes to the documentation~~

## How to test this PR locally

1. stack running 8.7.0 (SNAPSHOT ok)
2. build of elastic-agent reporting version 8.6.0 (can't be SNAPSHOT)
3. build of elastic-agent reporting version 8.7.0 (can't be SNAPSHOT)
4. binary artifact server where you can control download speed and hosts versions of elastic-agent
5. Add binary server to Fleet settings
6. Create policy that uses custom binary artifact server
7. Add 8.6.0 agent to policy that uses custom binary artifact server
8. Upgrade agent in fleet
9. notice checkins will continue
10. if download takes longer than 2 hours or is idle for at least 30sec, upgrade will error and agent will go back to "Healthy" and version 8.6.60
11. if download completes in under 2 hours, successful upgrade to 8.7.0

## Related issues

- Closes #1706
- Superseds #1666
